### PR TITLE
Add more specific GitHub issue templates

### PR DESCRIPTION
**Pull request by ianhi**
## PR Summary
Closes: #18160

- Added a feature request template that will automatically be tagged with the `New Feature` label. 
- Moved ISSUE_TEMPLATE.MD to ISSUE_TEMPLATE/bug_report.md so that both issue templates will show up as options for opening a new issue.

The feature request template is an almost exact copy of the jupyterlab template but I wrote the jupyterlab version so I think it's fine to just copy it w/o attribution. 

ping @story645 
## PR Checklist

- na Has Pytest style unit tests
- na Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- na New features are documented, with examples if plot related
- na Documentation is sphinx and numpydoc compliant
- na Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- na Documented in doc/api/next_api_changes/* if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by story645** [comment]
Thanks for opening this!  
For reference for other folk looking at this PR, went back into the MPL meeting notes and the discussion around this was to adopt the sklearn model (pandas does something similar/ it's kinda standard. These are what was agreed upon:

- [ ] Bug / Crash:  Something is happening that clearly shouldn’t happen. Wrong results as well as unexpected errors from estimators go here. 
- [ ] Cleanup / Enhancement:  Improving performance, usability, consistency. 
- [ ] Documentation:  Missing, incorrect or sub-standard documentations and examples. 
- [ ] New Feature: Feature requests and pull requests implementing a new feature.

**Comment by jklymak** [comment]
@story645 are you asking for all these categories to be implimented?  Whats the next step here?

**Comment by story645** [comment]
No, just the Bug and New Feature is great! I'm just putting this here cause we already agreed on it on a call and it's kind of a pain to pull up.

File: .github/ISSUE_TEMPLATE/feature_request.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+-->
+
+### Proposed Solution
+
+<!-- Provide a clear and concise description of a way to accomplish what you want. For example:
+
+* Add an option so that when [...]  [...] will happen
+ -->
+
+### Additional context
```
Comment (jklymak) [inline .github/ISSUE_TEMPLATE/feature_request.md:29]:
```suggestion
### Additional context and prior art
```

File: .github/ISSUE_TEMPLATE/feature_request.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
+---
+name: Feature Request
+about: Suggest something to add to Matplotlib
+labels: New Feature
+---
+
+<!--
+Welcome! Thanks for thinking of a way to improve Matplotlib. If this solves a problem for you, then it probably solves that problem for lots of people! So the whole community will benefit from this request.
```
Comment (jklymak) [inline .github/ISSUE_TEMPLATE/feature_request.md:8]: please fix the line breaks.... 

I'd just shorten to `Thanks for thinking of a way to improve Matplotlib.` and leave off the rest.

Comment (ianhi) [reply .github/ISSUE_TEMPLATE/feature_request.md:8]: > please fix the line breaks....

Do you mean inside the paragraph or do you mean how the text is positioned relative to to the comment lines?

Comment (jklymak) [reply .github/ISSUE_TEMPLATE/feature_request.md:8]: Just we wrap most lines at 79 characters.  Of course not as big a deal with a markdown file, but...

Comment (ianhi) [reply .github/ISSUE_TEMPLATE/feature_request.md:8]: got it. Would you be open to avoiding that rule in the issue templates? As they are prose rather than code I think it makes more sense for the github issue editor/browser to handle the reflow/wrapping. For example the line wrapping at 79 characters in the current issue template feels unnatural to me:
image

There are a few other places in the feature request template where I went over 79 lines, and the existing bug_report template seems to have inconsistent enforcement of this.

**Comment by ianhi** [comment]
I added a template for each of @story645's suggestions. I used the `Enhancement` label for the enhancement as that felt most natural but it does not seem to exist as a label currently. If there is a better label to use please let me know and I will change it.

**Comment by story645** [comment]
Thanks! We just discussed this on the call and decided on New Feature as the label for these sorts of things.

**Comment by dopplershift** [comment]
I don't want to bikeshed this PR to death, but to me "Enhancement" != "maintenance". You don't enhance your car by changing its oil, but it is maintenance. I'd say add the new label, but I'm content to be told how wrong I am. :wink:

**Comment by story645** [comment]
Ryan these are for clean ups & enhancements, which I think folks have been tagging maintenance? There's a seperate issue template for new features and we just discussed on the call how we don't want a standalone enhancement tag cause nobody knew what it was for.

File: .github/ISSUE_TEMPLATE/feature_request.md (hunk 2 above)
Comment (jklymak) [reply .github/ISSUE_TEMPLATE/feature_request.md]: I don't feel strongly about this, but others do. I think unless there is a good reason (i.e. it'll make horrible line continuations) we stick to 79 characters.

**Comment by jklymak** [comment]
I would not include the "Enhancement.md" at all.  We already have "Bug or issue" and "New Feature", and its hard to think of an "enhancement" request not covered under one of those two templates.

**Comment by story645** [comment]
From the [July 20th meeting notes](https://hackmd.io/MNcskNfZSIOoLb50GqVIGg)
>
 replace "enhancement" with "new feature"
- new feature makes new API
- performance enhancement & maintenance clean up code

& this would be for stuff tagged https://github.com/matplotlib/matplotlib/issues?q=is%3Aissue+is%3Aopen+label%3Amaintenance

**Comment by jklymak** [comment]
I'm confused about what you are trying to say @story645.  Are you arguing for keeping `enhancement.md` in this PR or removing it?

**Comment by story645** [comment]
>I'm confused about what you are trying to say @story645. Are you arguing for keeping enhancement.md in this PR or removing it?
keeping it as a template for maintenance type issues

File: .github/ISSUE_TEMPLATE/documentation.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+
+### Suggested Improvement
+
+<!--
+If you have an idea to improve the documentation please suggest it here
+
+* This line should be be changed to say [...]
+* Include a paragraph explaining [...]
+-->
```
Comment (story645) [inline .github/ISSUE_TEMPLATE/documentation.md:29]: * add a figure showing

**Comment by jklymak** [comment]
Which recent issues do you wish the user had used the proposed "Maintenance" template for?  Because I don't see the benefit from the user's point of view.

File: .github/ISSUE_TEMPLATE/enhancement.md [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+**Summary**
+
+<!-- A short 1-2 sentences that succinctly describe what could be improved -->
+
+**Code example**
+<!--A minimal code snippet required to reproduce the bug.
+Please make sure to minimize the number of dependencies required, and provide
+any necessary plotted data
+-->
+
```
Comment (story645) [inline .github/ISSUE_TEMPLATE/enhancement.md:18]: This isn't really a bug fix, this is a tweak/optimization/probably the why would we want to do this?

File: .github/ISSUE_TEMPLATE/feature_request.md [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+### Problem
+
+<!--
+Provide a clear and concise description of the problem this feature will solve. 
+
+For example:
+* I'm always frustrated when [...] because [...]
+* I would like it if [...] happened when I [...] because [...]
+-->
+
```
Comment (story645) [inline .github/ISSUE_TEMPLATE/feature_request.md:23]: * Here is a sample image of what I am asking for [...]

**Comment by ianhi** [comment]
Is there any harm in having 1~2 more issue templates? The difficulty here seems to be derived from the lumping of enhancement and cleanup together. How about splitting them up, giving a breakdown like this:

- **feature request** new api
- **enhancement** performance improvement or other small improvment
    - I often feel weird calling a request for small improvements a "feature request" so having a dedicated enhancement makes sense to me.
- **clean-up** correct inconsistencies


For an example of cleanup:
The acceptable usage of gitter is different on the matplotlib website:
image

vs on the readme:
image

**Comment by QuLogic** [comment]
Can we also add something like "I don't know what I want" that redirects to Discourse?

**Comment by story645** [comment]
@QuLogic some other projects make a "question/support/other" type template that explicitly says something like the issue tracker isn't for this sort of thing so we're most likely going to close this issue and you should use one of our other channels instead. 

Here's the scikit-learn version: 
https://www.github.com/scikit-learn/scikit-learn/tree/master/.github%2FISSUE_TEMPLATE%2Fusage_question.md

**Comment by ianhi** [comment]
So in summary the conclusion is to have the following:

- [x]  **feature request** new api
- [x]  **documentation** Missing, incorrect or sub-standard documentations and examples.
- [x] **enhancement** performance improvement or other small improvment
- [ ] **clean-up** correct inconsistencies
- [ ] **"question/support/other** redirect to Discourse

Is that a fair assement? If yes, shall I set the label for enhancement to `enhancement` and use the `maintenance` label for cleanup? I think this may require action by someone to actually create the label.

**Comment by QuLogic** [comment]
I think we might have crossed wires a bit. We _don't_ have an `enhancement` label and in fact, we deleted it recently as it seem `New feature` was more indicative of what it meant.

For performance and small improvements, we can lump that together with cleanup and label it `maintenance`. While we do have a separate `performance` label, the number of such PRs is much smaller, so I think it's not a great burden to add that label after the PR is opened. If it does turn out to be a burden, we can always add another template type later.

And to answer another question, *Enhancement* != *Maintenance*, but the difference here is a matter of scope. If it's an *Enhancement* that requires a What's new entry (maybe also for API notes too), then the *Enhancement* is a *New feature*. If it's an *Enhancement* that makes the library 'better' (for some definition of 'better'), but isn't user-visible, we'll call it *Maintenance*.

**Comment by QuLogic** [comment]
In terms of concrete to-do:
* merge enhancement and cleanup and call it **maintenance** (with that label), but try to avoid calling it an Enhancement
* add **question/support/other** redirect to Discourse

File: .github/ISSUE_TEMPLATE/questions.md [hunk 6]
```diff
@@ -0,0 +1,17 @@
+---
+name: Question/Support/Other
+about: If you have a usage question
+labels: Question
```
Comment (QuLogic) [inline .github/ISSUE_TEMPLATE/questions.md:4]:
```suggestion
labels: Community support
```

File: .github/ISSUE_TEMPLATE/feature_request.md [hunk 7]
```diff
@@ -0,0 +1,37 @@
+---
+name: Feature Request
+about: Suggest something to add to Matplotlib
+labels: New Feature
```
Comment (QuLogic) [inline .github/ISSUE_TEMPLATE/feature_request.md:4]:
```suggestion
labels: New feature
```

File: .github/ISSUE_TEMPLATE/maintenance.md [hunk 8]
```diff
@@ -0,0 +1,14 @@
+---
+name: Maintenance
+about: Help improve performance, usability and/or consistency.
+label: maintenance
```
Comment (QuLogic) [inline .github/ISSUE_TEMPLATE/maintenance.md:4]: (I renamed it capitalized for consistency.)
```suggestion
label: Maintenance
```

File: .github/ISSUE_TEMPLATE/questions.md [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+# This template is based on the scikit-learn template
+# https://github.com/scikit-learn/scikit-learn/blob/8e61534f1087703f476414d8dbd3688282f8eebf/.github/ISSUE_TEMPLATE/usage_question.md
+---
+
+<!--
+**If your issue is a usage question, please submit it in one of these other channels instead:**
+- StackOverflow with the Matplotlib tag: https://stackoverflow.com/questions/tagged/matplotlib
+- The Matplotlib community forum: https://discourse.matplotlib.org/
```
Comment (story645) [inline .github/ISSUE_TEMPLATE/questions.md:12]: this one should come first since it's the one we prioritize.

Comment (ianhi) [reply .github/ISSUE_TEMPLATE/questions.md:11]: Makes sense. I swapped the order.

File: .github/ISSUE_TEMPLATE/documentation.md [hunk 10]
```diff
@@ -0,0 +1,30 @@
+
+---
+name: Documentation improvement
+about: Create a report to help us improve the documentation.
```
Comment (QuLogic) [inline .github/ISSUE_TEMPLATE/documentation.md:4]: This is the only one with a period.
```suggestion
about: Create a report to help us improve the documentation
```

**Comment by ianhi** [comment]
> This is the only one with a period.

not only that but there was an extra line at the top of the file that I believe would have caused github to not pick up the template. I think they all should be good now

File: .github/ISSUE_TEMPLATE/questions.md [hunk 11]
```diff
@@ -0,0 +1,17 @@
+---
+name: Question/Support/Other
+about: If you have a usage question
+labels: Community support
+# This template is based on the scikit-learn template
+# https://github.com/scikit-learn/scikit-learn/blob/8e61534f1087703f476414d8dbd3688282f8eebf/.github/ISSUE_TEMPLATE/usage_question.md
```
Comment (story645) [inline .github/ISSUE_TEMPLATE/questions.md:6]: Going back and forth on whether this needs to be in here...

Comment (ianhi) [reply .github/ISSUE_TEMPLATE/questions.md:6]: It won't show up when someone makes a new issue as it is a comment in the yaml frontmatter rather than in the markdown body. See example here: https://github.com/ianhi/git-expt/issues/new?labels=Question&template=test.md and the source file https://raw.githubusercontent.com/ianhi/git-expt/master/.github/ISSUE_TEMPLATE/test.md

Comment (story645) [reply .github/ISSUE_TEMPLATE/questions.md:6]: thanks for the clarification.

**Comment by story645** [comment]
Thank you @ianhi for knocking something off my to do list 😉!

**Comment by ianhi** [comment]
Thanks @story645 for helping get it through! I really appreciate having nice templates when I open an issue
